video: introduction of STM32 VENC driver for H264 hardware video compression

hfruchet-st opened the pull request:
This PR introduces H264 hardware video compression thanks to VENC peripheral of STM32N6.

A new stm32 video encoder driver has been introduced which relies on vc8000nanoe software stack [1].
This is a first basic porting to enable H264 encoding use-case.
RAM memory resources required by VENC encoding process are allocated in external PSRAM, camera frames
captured by DCMIPP and encoded by VENC are also allocated in external PSRAM.
The video encoder driver currently lack of controls to configure the encoding process, default configuration
targets H264 1080p 2Mb/s constant bitrate.

In order to test, the tcpserversink video sample has been enhanced to support video compression in order to stream small H264 compressed video chunks instead of big raw uncompressed images [2].

[1] https://github.com/zephyrproject-rtos/hal_stm32/pull/295
[2] https://github.com/zephyrproject-rtos/zephyr/pull/95862

hfruchet-st committed:
drivers: video: introduction of the stm32 venc driver

The STM32 video encoder (VENC) peripheral is a hardware
accelerator allowing to compress RGB/YUV frames into
H264 video bitstream chunks.

Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>

hfruchet-st committed:
dts-bindings: video: addition of stm32 venc description

Addition of description for the STM32 Video encoder (VENC).

Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>

hfruchet-st committed:
dts: arm: st: n6: add venc node

Add node describing the venc in stm32n6.dtsi

Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+LOG_MODULE_REGISTER(video_stm32_venc, CONFIG_VIDEO_LOG_LEVEL);
+
+#define EWL_HEAP_ALIGNED_ALLOC(size)				\
+	mem_attr_heap_aligned_alloc((ATTR_SW_ALLOC_NON_CACHE <<	\
+				    DT_MEM_SW_ATTR_SHIFT), ALIGNMENT_INCR, size)
```
```suggestion
	mem_attr_heap_aligned_alloc(DT_MEM_SW_ALLOC_NON_CACHE, ALIGNMENT_INCR, size)
```

Also, this uses `ALIGNMENT_INCR` which is defined later in this file.
(I know this will work but it's confusing to read)

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+				    DT_MEM_SW_ATTR_SHIFT), ALIGNMENT_INCR, size)
+#define EWL_HEAP_ALIGNED_FREE(block) mem_attr_heap_free(block)
+
+#define VENC_RATECTRL_FPS 30
+#define VENC_RATECTRL_QP 25
```
Nit: if those are user-configurable, they could be promoted to Kconfig symbols.

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+  i32 irq_cnt;
+} VENC_EWL_TypeDef;
+
+#define ALIGNED_SIZE(x) ((((x) + ALIGNMENT_INCR - 1UL) / ALIGNMENT_INCR) * ALIGNMENT_INCR)
```
Use the Zephyr implementation (from `#include <zephyr/sys/util.h>`)
```suggestion
#define ALIGNED_SIZE(x) ROUND_UP(x, ALIGNMENT_INCR)
```
(or use a raw `ROUND_UP` everywhere in place of `ALIGNED_SIZE` - as you prefer)

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+
+#define ALIGNED_SIZE(x) ((((x) + ALIGNMENT_INCR - 1UL) / ALIGNMENT_INCR) * ALIGNMENT_INCR)
+
+static VENC_EWL_TypeDef ewl_instance;
```
IMO, data definitions should come after declarations for clarity (i.e., move to L97)

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+	cfg_info.searchAreaSmall = (cfgval >> 29U) & 1U;
+	cfg_info.scalingEnabled = (cfgval >> 30U) & 1U;
+
+	cfg_info.busType = (cfgval >> 20U) & 15U;
+	cfg_info.synthesisLanguage = (cfgval >> 16U) & 15U;
+	cfg_info.busWidth = (cfgval >> 12U) & 15U;
```
Nit: IMO, prefer hex notation (or use `GENMASK`) for masks with more than 1 bit.
```suggestion
	cfg_info.busType = (cfgval >> 20U) & 0xFU;
	cfg_info.synthesisLanguage = (cfgval >> 16U) & 0xFU;
	cfg_info.busWidth = (cfgval >> 12U) & 0xFU;
```

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+	k_sem_init(&ewl_instance.complete, 0, 1);
+	k_sem_reset(&ewl_instance.complete);
+
+	/* set client type */
+	ewl_instance.clientType = param->clientType;
+
+	mem_attr_heap_pool_init();
+
+	ewl_instance.irq_cnt = 0;
```
Nit: IMO, avoid mixing Zephyr and HW code
```suggestion
	/* set client type */
	ewl_instance.clientType = param->clientType;

	ewl_instance.irq_cnt = 0;

	mem_attr_heap_pool_init();
```
(I would even drop newline between the two top lines, but I kept it for the comment to be more scoped)

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+		__func__, offset / 4, offset, val);*/
+
+	/* offset is defined in the EWL API in bytes. convert to number of registers by shifting */
+	_LL_VENC_WriteRegister((offset >> 2), val);
```
Nit: I would integrate the `offset >> 2` inside `_LL_VENC_WriteRegister`.
(A non-LL branded name, e.g. `venc_write_reg()` would then be more sensical)

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+			break;
+		}
+
+		if (slicesReady)
+			if (*slicesReady > prevSlicesReady) {
+				ret = EWL_OK;
+				break;
+			}
+
```
```suggestion
		if (slicesReady) {
			if (*slicesReady > prevSlicesReady) {
				ret = EWL_OK;
				break;
			}
		}
```
or combine in same condition: `if (slicesReady != NULL && *slicesReady > prevSlicesReady)`

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+
+	LOG_DBG("EWLWaitHw: encoding = %d ms\n", k_ticks_to_ms_ceil32(sys_clock_tick_get_32() - start));
+
+	if (slicesReady != NULL)
+		LOG_DBG("EWLWaitHw: slicesReady = %d\n", *slicesReady);
```
```suggestion
	if (slicesReady != NULL) {
		LOG_DBG("EWLWaitHw: slicesReady = %d\n", *slicesReady);
	}
```

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
```
No calls to `clock_control_on`?

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+{
+	/* VENC HW REset */
+	__HAL_RCC_VENC_FORCE_RESET();
+	k_sleep(K_MSEC(1));
```
Maybe `k_busy_wait` with a lower delay (if upper bound for `tRESET` is known and <<< 1ms).

Although polling a "is IP under reset" bit would be better. (Ideally, we would even call `reset_line_toggle` with info from DT, but driver doesn't perform polling...)

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+
+	if (!data->encoder) {
+		ret = encoder_prepare(data);
+		if (ret)
+			return -1;
```
```suggestion
		if (ret) {
			return -1;
		}
```
or even `ret < 0` and propagate `ret`

hfruchet-st replied:
This is not same coding style than Linux kernel, is there a doc pointing out the coding style to match ? Or better: a script like Linux kernel "checkpatch" ?

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+	}
+	LOG_DBG("output=%p\n", output);
+
+	if (!(data->frame_nb % 30) || data->resync) {
```
If this `30` is framerate dependent, it should come from a macro.

hfruchet-st replied:
This first implementation only supports fixed 30fps encoding.

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+ISR_DIRECT_DECLARE(stm32_venc_isr)
+{
+	VENC_EWL_TypeDef *enc = &ewl_instance;
+	u32 hw_handshake_status = READ_BIT(VENC_REG(BASE_HEncInstantInput >> 2U), (1U << 29U));
```
No HW define for (1 << 29)? (Nit: should be `1U << 29`)

josuah commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+
+	output = k_fifo_get(&data->fifo_output_in, K_NO_WAIT);
+	if (!output) {
+		return 0;
```
At this line, the `input` buffer is dequeued and dropped by the driver.

There would need some scheme to synchronize both queue to make sure to have both an input and an output buffer poped from the FIFO at the same time.

- One method is to turn it into a thread with `K_FOREVER` instead of `K_NO_WAIT`, and therefre, never `return 0` as they are guaranteed to return something with `K_FOREVER`.

- One method is to 'peek' at the FIFO without popping the buffer. If there is a buffer ready on both FIFO, then go ahead and pop them on both in confidence that none of them would be NULL (as it was checked beforehand)...

Maybe this does not cause an issue in the way the `tcpserversink` uses this driver though!

josuah reviewed (commented):
Thank you for preparing VENC!

This will really make [video over network on Zephyr](https://github.com/zephyrproject-rtos/zephyr/discussions/92843#discussioncomment-13743497) interesting.

It seems like Zephyr lacks a few APIs from Linux (no H.264 format definition, no planar API for NV12...), so if anything feels like missing, let us know so that the missing API can be provided in parallel.

avolmat-st commented on include/zephyr/drivers/video.h:
```diff
@@ -1597,6 +1599,7 @@ static inline unsigned int video_bits_per_pixel(uint32_t pixfmt)
 	case VIDEO_PIX_FMT_SGRBG12P:
 	case VIDEO_PIX_FMT_SRGGB12P:
 	case VIDEO_PIX_FMT_Y12P:
+	case VIDEO_PIX_FMT_NV12:
```
I've added the whole multi buffer support of DCMIPP as well as description in video.h in the following PR which is still currently in draft so that I can fully test it:

https://github.com/zephyrproject-rtos/zephyr/pull/94081

erwango commented on samples/drivers/video/tcpserversink/Kconfig:
```diff
@@ -0,0 +1,63 @@
+# Copyright (c) 2024 Espressif Systems (Shanghai) Co., Ltd.
```
Use your own copyright.

erwango commented on samples/drivers/video/tcpserversink/Kconfig:
```diff
@@ -48,6 +48,15 @@ config VIDEO_PIXEL_FORMAT
 
+config VIDEO_NUM_FRAMES
+	int "Capture N-buffering"
+	default 0
```
```
	default 2
```
to fit the description

erwango commented on samples/drivers/video/tcpserversink/src/main.c:
```diff
@@ -135,7 +135,11 @@ int main(void)
 {
 	struct sockaddr_in addr, client_addr;
 	socklen_t client_addr_len = sizeof(client_addr);
+#if CONFIG_VIDEO_NUM_FRAMES
+	struct video_buffer *buffers[CONFIG_VIDEO_NUM_FRAMES];
+#else
 	struct video_buffer *buffers[2];
+#endif
```
Drop the #if/#else

erwango commented on samples/drivers/video/tcpserversink/README.rst:
```diff
@@ -26,6 +29,10 @@ J35 camera connector. A USB cable should be connected from a host to the micro
+On :zephyr:board:`stm32n6570_dk`, the MB1854 IMX335 camera module must be plugged in
+the CSI-2 camera connector. A RJ45 ethernet cable must be plugged in the ethernet CN6
+connector.
+
```
Probably add a note here mentioning the ISP module should be used as it cannot be added to the default west manifest.

erwango commented on boards/st/stm32n6570_dk/stm32n6570_dk_common.dtsi:
```diff
@@ -378,7 +378,7 @@ zephyr_udc0: &usbotg_hs1 {
 };
 
 &mac {
```
Drop this change. If required add overlay in the sample

erwango reviewed (changes requested):
First round of comment, driver aside.
Please split the PR and do a specific PR for the first commits that are not directly linked to the VENC driver. These can be reviewed and integrated faster.
For the changes on the sample, I let video team decide if this should be kept within the tcpserver sample or split into a different sample

erwango commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+	EWLHwConfig_t cfg_info;
+	u32 cfgval;
+
+	LOG_DBG("%s>\n", __func__);
```
LOG_DBG already provides the func IIRC

thecapn32 committed:
include: driver: video: add h264 pixel format support

Add H264 pixel format support.

Signed-off-by: Mohammad Massoudi <m.massoudi32@gmail.com>

erwango reviewed (commented):
Some pending comments, not sure why they were missing

hfruchet-st commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,1026 @@
+
+	VENC_EWL_TypeDef *inst = (VENC_EWL_TypeDef *) instance;
+	/* make size 8-byte aligned */
+	u32 size_aligned = ALIGNED_SIZE(size);
```
all u32 changed to uint32_t

avolmat-st commented on drivers/video/Kconfig.stm32_venc:
```diff
@@ -0,0 +1,14 @@
+	depends on DT_HAS_ST_STM32_VENC_ENABLED
+	select HAS_STM32LIB
+	select USE_STM32_LL_VENC
+	select USE_STM32_HAL_RIF if SOC_SERIES_STM32N6X
```
Actually the driver doesn't do reset so it should either be added in the driver (and keeping this select) or avoid putting it here and removing it from the binding/dts.

avolmat-st commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,894 @@
+	output->bytesused = encOut.streamSize;
+	LOG_DBG("SPS/PPS generated, size= %d", output->bytesused);
+
+	k_fifo_put(&data->fifo_output_in, output);
```
My understanding is that at start time the encoder needs 1 output buffer which will be later on pushed back to the application. For that purpose, the buffer is pulled out of the output_in fifo at the beginning of the function.
However since it is pushed back to the applicatiob at the end via the fifo_output_out fifo, it shouldn't be pushed into the fifo_output_in as well.

avolmat-st commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,894 @@
+{
+	struct video_stm32_venc_data *data = dev->data;
+
+	*vbuf = k_fifo_get(&data->fifo_output_out, timeout);
```
From a pipeline management point of view, the "element" making the connection between the camera pipeline and the encoder would expect to be able to do usual queue/dequeue of buffers and perform the exchange between the encoder and the camera pipeline so I think it is still important to have the proper fifo_input_in / fifo_input_out to be able to give back input buffer to the application.

josuah commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,894 @@
+
+	/* VENC produces full frames */
+	caps->min_line_count = caps->max_line_count = LINE_COUNT_HEIGHT;
+
```
The DCMI driver was missing this too which might have been misleading
- https://github.com/zephyrproject-rtos/zephyr/pull/96076

hfruchet-st reviewed (commented):
Thank you for your comments, bunch of fixes sent related to comments and also to QA tools.
Still some opened points on which I have replied...
Still some comments to process...

JarmouniA reviewed (dismissed):
> Zephyr EWL wrapping layer code is copied/pasted from EWL reference code that can be found here: https://github.com/STMicroelectronics/STM32CubeN6/blob/main/Middlewares/ST/VideoEncoder_EWL/ewl_impl.c
I don't want to deviate too much from this sensible code, this will make the potential bug fixing reporting to this code much harder.

Code integrated in Zephyr shall follow Zephyr coding guidelines and best practices, use Zephyr APIs, recommended MACROs.... 
If there is a need to not modify it, it shall be hosted in the HAL repo instead.

hfruchet-st commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,886 @@
+	__ASSERT_NO_MSG(inst != NULL);
+}
+
+void *EWLmalloc(uint32_t n)
```
OK, sync issue with EWL ref code, fixed.

hfruchet-st commented on doc/build/dts/api/api.rst:
```diff
@@ -466,6 +466,8 @@ device.
    * - zephyr,touch
      - touchscreen controller device node.
+   * - zephyr,videoenc
+     - Video encoder device, typically an H264 or MJPEG video encoder.
```
Because venc node activation as been moved to tcpserversink sample PR, so let's move also this documentation... See https://github.com/zephyrproject-rtos/zephyr/pull/95862/commits/a4d5efc7c98bcb2322704bda8930b898de7529a7

mathieuchopstm reviewed (commented):
One nit, but otherwise LGTM (though I haven't reviewed every single detail).

JarmouniA reviewed (commented):
The commit pulling hal_stm32 PR should be the 1st.

JarmouniA reviewed (commented):
SonarQube issues that need attention
https://sonarcloud.io/project/issues?id=zephyrproject-rtos_zephyr&pullRequest=92884&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true

mathieuchopstm commented:
> SonarQube issues that need attention https://sonarcloud.io/project/issues?id=zephyrproject-rtos_zephyr&pullRequest=92884&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true

The first two are incorrect because `instance` is really `&ewl_instance` returned by:
https://github.com/zephyrproject-rtos/zephyr/blob/a0f000d9f875193cb4b3b822691aca5eaebf6386/drivers/video/video_stm32_venc.c#L137-L151

The third one is the opposite: API takes an opaque `void*` which the consumer (RCC driver) never writes to, so dropping the `const` qualified is not incorrect.

etienne-lms commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,855 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_venc
```
Ok, my review comment can be addressed later, if still applicable.

That said, there is no effective need to have this macro defined before header file inclusion AFAICT. It's not mentioned in the [doc](https://docs.zephyrproject.org/latest/build/dts/howtos.html#option-1-create-devices-using-instance-numbers). Do you think it should be?

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,855 @@
+	case VIDEO_PIX_FMT_RGB565:
+		return H264ENC_RGB565;
+	default:
+		CODE_UNREACHABLE;
```
No, this should definitely be addressed now if reachable. `CODE_UNREACHABLE` can trash the codegen and cause problems.

mathieuchopstm commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,855 @@
+	return sys_read32(config->reg + BASE_HEncASIC);
+}
+
+EWLHwConfig_t EWLReadAsicConfig(void)
```
> It is really nice to return a full structure here? Couldn't the API provides a pointer this function could fill?

FYI this will be done automatically under the hood as it is required by the ARM calling convention :slightly_smiling_face:
(codegen should be mostly identical c.f. https://godbolt.org/z/3bqqPGeWs)

hfruchet-st committed:
include: video: add size field to video_format structure

Add size field to the video_format structure which needs to be set by the
driver and exposed to the application.

For uncompressed formats, this is the size of the raw data buffer in
bytes, which could be the whole raw image or a portion of the raw image
in cases the receiver / dma supports it.

For compressed formats, this is the estimated maximum number of bytes
required to hold a complete compressed frame.

Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

erwango committed:
samples: video: tcpserversink: Add stm32n6570_dk as integration platform

Required to build VIDEO_STM32_VENC.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>
Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>

mathieuchopstm commented on dts/bindings/video/st,stm32-venc.yaml:
```diff
@@ -0,0 +1,23 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: |
+    STMicroelectronics STM32 video encoder peripheral (VENC).
+    Example of node configuration at board level:
```
Stray line remaining
```suggestion
description: |
    STMicroelectronics STM32 video encoder peripheral (VENC).
```

But can simplify to:
```suggestion
description: STM32 video encoder (VENC)
```

mathieuchopstm reviewed (commented):
Otherwise LGTM.

etienne-lms reviewed (commented):
Thanks for the updates.

etienne-lms commented:
By the way, *"twister-build (3)"* CI test failure is unrelated and should be fixed by https://github.com/zephyrproject-rtos/zephyr/pull/96859. @hfruchet-st, you'll need to rebase.

thecapn32 commented:
> @thecapn32, WOuld you be okay to provide @hfruchet-st a real name he can use in the commit message?

Sure I have added my name to my profile.

thecapn32 commented:
Ah Alright then.
Name: Mohammad Massoudi
Email: m.massoudi32@gmail.com

avolmat-st commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,889 @@
+		ret = encoder_start(data, output);
+
+		/* Output buffer is now filled with SPS/PPS header, return it to application */
+		goto out;
```
My understanding is that when starting the encoder, this will only produce data into the output buffer, and input buffer is not used in such case. Is that correct ?

I feel that in the way it is currently done, the first input frame is going to be lost / unused and returned to the application (and thus back to the camera side) without actually having been used.

If that's the case it might be better to add a comment here, with maybe a TODO / FIXME or something like  that in order to address in a further PR.

avolmat-st commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,889 @@
+	LOG_DBG("output=%p, encOut.streamSize=%d", output, encOut.streamSize);
+
+	k_fifo_put(&data->out_fifo_in, output);
+	k_fifo_put(&data->out_fifo_out, output);
```
The same output cannot be put back into both side of the output queue. Doing so would mean that for example the encoder could override the output frame while it is currently used by the application (currently tcpserversink), is my understanding correct ? It should only be put into the out_fifo_out I think.

hfruchet-st replied:
My bad, lines removed.

avolmat-st commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,889 @@
+
+out:
+	k_fifo_put(&data->in_fifo_out, input);
+	k_fifo_put(&data->out_fifo_out, output);
```
doing k_fifo_put of output into the out_fifo_out leads to pushing the output buffer twice into out_fifo_out if we consider the k_fifo_put already done above ??

avolmat-st commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,889 @@
+
+static int stm32_venc_get_caps(const struct device *dev, struct video_caps *caps)
+{
+	caps->format_caps = fmts;
```
I think the format_caps should depends on the caps->type and return different format depending on INPUT or OUTPUT. Currently that might be RGB565/NV12 for the INPUT and H264 for the OUTPUT ?

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -79,6 +79,12 @@ struct video_format {
+	 *
+	 * This is the maximum size in bytes required for buffer data.
+	 */
+	uint32_t sizeimage;
```
Thanks for this adding. I think the field name `sizeimage` may be due to some historical reason from V4L1. Should it be better to rename it to simply `size` and add some more documentation, something like

```
	/**
	 * @brief Size of the image in bytes
	 *
	 * For uncompressed formats, it is the size of the raw image in bytes.
	 * 
	 * For compressed format, this is the maximum number of bytes
	 * required to hold a complete compressed frame.
	 */
	uint32_t size;
```

ngphibang replied:
Yes, maybe reworded to 

`For uncompressed formats, it is the size of the raw data buffer in bytes.` ?

buffer here means a whole raw image or a portion of image (for e.g. smartdma)

hfruchet-st commented:
Rebased.

avolmat-st reviewed (dismissed):
Looks good to me. Thanks.

erwango reviewed (changes requested):
Otherwise LGTM

hfruchet-st commented on west.yml:
```diff
@@ -250,7 +250,7 @@ manifest:
     - name: hal_stm32
-      revision: b6f03b6efd92da4564b38df8cb54d7dad5f37163
+      revision: 18186afb8d08cf1c2af31b656636617afcebad0c
       path: modules/hal/stm32
```
Done

josuah commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,899 @@
+			return -EINVAL;
+		}
+
+		fmt->size = fmt->width * fmt->height / VENC_ESTIMATED_COMPRESSION_RATIO;
```
A compression ratio would be comparing the input size (`fmt->pitch` rather than `fmt->width`) to the output size (in H.264), I suppose here all pixel formats (RGB565, NV12) are first converted to a same format, then compressed?

Otherwise the output size (bytesused) and compression ratio would be different for NV12 and RGB565?

Not blocking and I was just curious about it.
Thanks!

josuah replied:
Very good to know. This makes sense the way it is and having fine estimation are not blockers of course.
Thank you.

josuah commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,899 @@
+
+	encIn.pOutBuf = (uint32_t *)output->buffer;
+	encIn.busOutBuf = (uint32_t)encIn.pOutBuf;
+	encIn.outBufSize = ROUND_UP(output->size, ALIGNMENT_INCR);
```
Should this be capped to the `fmt->size` announced? As it is said to be the maximum size by `video.h`.

Which semantics would you wish for fmt->size? If it is the max size, then `fmt->max_size` would be fit? I will follow-up on https://github.com/zephyrproject-rtos/zephyr/pull/92884#discussion_r2409617631

josuah commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,899 @@
+	H264EncIn encIn = {0};
+	H264EncOut encOut = {0};
+
+	if (k_fifo_is_empty(&data->in_fifo_in) || k_fifo_is_empty(&data->out_fifo_in)) {
+		/* Encoding deferred to next buffer queueing */
+		return 0;
+	}
+
+	input = k_fifo_get(&data->in_fifo_in, K_NO_WAIT);
+	output = k_fifo_get(&data->out_fifo_in, K_NO_WAIT);
```
There is also poor documentation on what guarantees video drivers can expect from to `video_enqeue()` regarding concurrency. A known quirk to fix in video area.

- https://github.com/zephyrproject-rtos/zephyr/issues/89109

This mean that in the meantime, there is a chance that `video_enqueue()` can be called from 2 different threads at the same time, as driver APIs are currently called directly without lock.

We could end-up with `input` and/or `output` being NULL in one of the thread.

Would it make sense to wrap this block in a `mutex_lock()` in the meantime that the buffering improvements are complete?

- https://github.com/zephyrproject-rtos/zephyr/pull/92370

hfruchet-st replied:
Locking added on _queue()/_dequeue().

josuah commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,899 @@
+		*vbuf = k_fifo_get(&data->in_fifo_out, timeout);
+	} else {
+		*vbuf = k_fifo_get(&data->out_fifo_out, timeout);
+
```
Some left-over empty line, just in case you wanted this cleared out...

josuah commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,899 @@
+	return 0;
+}
+
+ISR_DIRECT_DECLARE(stm32_venc_isr)
```
Just for curiosity: is this performance-critical since it uses an un-wrapped ISR?

josuah commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,899 @@
+	/* Define and initialize the master configuration structure */
+	RIMC_MasterConfig_t RIMC_master = {0};
+
+	/* Enable the clock for the RIFSC (RIF Security Controller) */
+	__HAL_RCC_RIFSC_CLK_ENABLE();
```
Sorry I do not know the STM32N6 hardware very well. Out of curiosity, is this unrelated to this?
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/reset/st%2Cstm32-rcc-rctl.yaml

josuah replied:
Understood, good to learn about the connection.

josuah reviewed (commented):
I prefer to give you time to answer before giving a +1.
In particular for the new API introduced.

This could go in the [release note](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-4.3.rst#new-apis-and-options) too, which has a "new API and options" section.

Thanks for bringing the first h.264 driver to Zephyr!

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -79,6 +79,15 @@ struct video_format {
+	/**
+	 * @brief size of the buffer in bytes.
+	 *
+	 * For uncompressed formats, this is the size of the raw image in bytes.
```
> After this change, for every driver currently in-tree:
> 
> The pitch is 0 for every compressed format, and >0 for uncompressed formats.
> The size is >0 for every compressed format, and 0 for uncompressed formats.
> For video receivers drivers, this mean the size is not matching this new rule as it is 0 instead of pitch * height?

I think after this change, application could base only on fmt.size to allocate buffers. The drivers need to set fmt.size (and the optional fmt.pitch in case we need it in the future).

min_line_count / max_line_count does not need to be exposed anymore (can be dropped in caps).

ngphibang replied:
`However, size would be more fmt->pitch * caps->max_line_count if max_line_count != LINE_COUNT_HEIGHT or fmt->pitch * fmt->height. Right ?`

`fmt->size = fmt->pitch * MIN_LINE_COUNT;`

and MIN_LINE_COUNT should be just defined inside some drivers that do not support full frame like smartdma.
Drivers should expose the fmt->size (and optionally fmt->pitch if required in the future, for now, it's not used).

Dropping the caps's min_line_count / max_line_count is good because it's only needed for some drivers but enforce all drivers to set it to -1. Also, application usually doesn't know what it is.

ngphibang replied:
> Yes, agree on the removal of those, I was thinking about how to make the transition to the usage of fmt->size while not all drivers already set it themself. Aka what would have to be the "glue" code during the transition.

I don't fully understand. Do we just change the sample code:

```
	/* Size to allocate for each buffer */
	if (caps.min_line_count == LINE_COUNT_HEIGHT) {
		bsize = fmt.pitch * fmt.height;
	} else {
		bsize = fmt.pitch * caps.min_line_count;
	}
```

to

`bsize = fmt.size`

And make "receiver" drivers to set fmt.size inside set/get_fmt() as we did with fmt.pitch ?

ngphibang replied:
> Introduce a MIN_LINE_COUNT video control only for devices having restrictions

Yes, nearly. MIN_LINE_COUNT is inside each driver, not system-wide, just like a macro SDMA_LINE_COUNT.

> Add a line at the end of video_set_format() to auto-set fmt.size to provide a default.

And get_format() too.

ngphibang replied:
Here is the PR https://github.com/zephyrproject-rtos/zephyr/pull/97197

josuah replied:
Good idea, better solution, and the PR in question should be mergeable relatively quickly.

josuah replied:
That sounds like a bit of an extra mile to go but also mean a lot fewer discussion on that thread and gets the problem faced out of the way.

josuah commented on drivers/video/video_stm32_venc.c:
```diff
@@ -0,0 +1,899 @@
+	case VIDEO_PIX_FMT_RGB565:
+		return H264ENC_RGB565;
+	default:
+		__ASSERT_NO_MSG(false);
```
Thank you I missed that.

hfruchet-st committed:
doc: release-note-4.3: document size field added to video_format struct

Document introduction of the new size field inside video_format struct
as visible on the <zephyr/drivers/video.h> header.

Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>

avolmat-st reviewed (dismissed):
LGTM. Thanks !!

josuah reviewed (dismissed):
Keeping the DNM flag to give a chance for this driver being added to CI (the snippet being merged), while #97197 keeps going in parallel.

josuah reviewed (dismissed):
Quoting https://github.com/zephyrproject-rtos/zephyr/pull/92884#issuecomment-3384814279

> Otherwise, I can make an enhancement commit and rebase https://github.com/zephyrproject-rtos/zephyr/pull/97197 if this one is merged first.

So looks like everybody is aligned on merging this now.

etienne-lms reviewed (approved):
Sorry for the late nitpicking comments, none are blocking. I'm fine getting the current series merged.
